README.md
# YANG Catalog

## Overview

This repository contains configurations and scripts that tie all the individual components of YANG Catalog together. The components themselves are included as git submodules.

You can find the official YANG Catalog website [here](https://yangcatalog.org).

For accessing some endpoints of YANG Catalog's API directly, you can use [ycclient](https://github.com/earies/ycclient).

# Deployment
The main repository to start up all the pieces of YANG Catalog

## Requirements

* Docker, including docker-compose
* IPv4 Internet connectivity (Docker Hub doesn't do IPv6 yet)
* Sufficient space for container images (approx 10 GB)
* Sufficient space for all the module data
  - approx 45 GB for local development
  - up to 250 GB for production
* Sufficient RAM
  - approx 10 GB for local development
  - up to 24 GB for production

### External requirements

The deployment uses several services via container images that are
distributed by third parties, e.g. via DockerHub

* OpenSearch
* RabbitMQ
* Redis
* NGINX (as the base image of the frontend container, which includes static content)

## Basic Usage

Create `conf/yangcatalog.conf` from `conf/yangcatalog.conf.sample`,
replacing values enclosed in `@...@` with your own.

```
git submodule update --init
docker-compose build # not strictly necessary
docker-compose up
```

The `docker-compose build` command will build container images for the
various components of the YANG Catalog.

The `docker-compose up` will start containers from these images, as
well as the necessary third-party images and combine them into a functional local deployment
of the YANG Catalog, which should be accessible at
http://localhost

All containers have access to a shared volume at `/var/yang`, which is also mapped to the same directory on the host.

### OpenSearch

For development, OpenSearch is run directly in the local environment as a docker container.
The `opensearch-aws` option needs to be set to `False`, `opensearch-host`,
which is the hostname containers will use to try and connect to OpenSearch,
needs to be set to `yc-opensearch`, and `opensearch-port` to `9200`.

In the production environment, we use [Amazon's OpenSearch Service](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/what-is.html).
Here `opensearch-aws` needs to be set to `True`, `opensearch-host` and `opensearch-port` must be configured
according to the information provided in the AWS console.


### Logging

All logs are stored in a directory configured in yangcatalog.conf, by default `/var/yang/logs`.
Logs can also be read and filtered through YANG Catalog's admin UI.

### ConfD

YANG Catalog requires ConfD, which is proprietary software and can not be freely distributed.
ConfD can be aquired from [Tail-F Systems](https://www.tail-f.com).

### Kubernetes support

It is possible to deploy YANG Catalog on Kubernetes by following this [guide](./k8s/README.md).

### OpenShift support

OpenShift is, at its core, a multi-tenant variant of Kubernetes.  It
imposes some limitations on the containers it runs; for example,
processes run under a non-root UID.  Many container images are build
under the assumption that they _will_ run as root.

## Structure of the Repository

Various open-source components of YANG Catalog are imported as
submodules:

* [backend](https://github.com/YangCatalog/backend) - YANG Catalog
  API server
* [frontend](https://github.com/YangCatalog/frontend) - Angular applications for 
  YANG Catalog frontend and admin UI together with some static content
* [yangre-gui](https://github.com/plewyllie/yangre-gui) - Peter
  Lewyllie's YANG Regular Expression checker
* [yangvalidator](https://github.com/YangCatalog/yang-validator-extractor) - Carl
  Moberg's YANG validation application
* [module-compilation](https://github.com/YangCatalog/module-compilation) - Scripts
  to analyze and validate YANG files

`docker-compose.yml` is the actual "orchestration" that attempts to
describe a complete (modulo ConfD) deployment of YANG Catalog.

The `conf` directory contains a few files that are passed to the
containers using bind mounts.

### .env variables

Some of the variables here have to correspond with the yangcatalog.conf file.

`COMPOSE_PROJECT_NAME=yc` - The prefix docker-compose will add to container names belonging to this project.
For example, `yc` will result in containers names `yc-backend`, `yc-frontend`, etc..

`REDIS_VOLUME` - Directory where redis will store its `dump.rdb` file.

`RABBITMQ_USER=<RABBITMQ_USER>` - rabbitmq username.

`RABBITMQ_PASSWORD=<RABBITMQ PASSWORD>`  - rabbitmq password.

`OPENSEARCH_DATA=/var/lib/opensearch` - If local OpenSearch is used this
variable is used to specify where OpenSearch indexed data should be saved

`OPENSEARCH_LOG=/var/logs/opensearch`  - If local OpenSearch is used this
variable is used to specify where OpenSearch logs should be saved

`NGINX_LOG=/var/logs/nginx` - Where do we save nginx logs

`KEY_FILE=/home/yang/deployment/resources/yangcatalog.org.key` - certificate key file for HTTPS protocol.
If None exists just use any path

`CERT_FILE=/home/yang/deployment/resources/yangcatalog.org.crt` - certificate file for HTTPS protocol.
If None exists just use any path

`CA_CERT_FILE=/home/yang/deployment/resources/yangcatalog.org.crt` - ca_certificate file for HTTPS protocol.
If None exists just use any path

`OPENSEARCH_ID=1016` - If local OpenSearch is used this
variable is used to set permissions to specific user ID. It`s safe to use same ID as YANG_ID

`OPENSEARCH_GID=1001` - If local OpenSearch is used this
variable is used to set permissions to specific group ID. It`s safe to use same ID as YANG_GID

`YANG_ID=1016` - ID created for yang user

`YANG_GID=1001` - GID created for yang user

`YANG_RESOURCES=/var/yang` - specify path where all shared YANG Catalog data will be saved.

`NGINX_FILES=nginx-testing.conf` - NGINX config files used for NGINX. There is testing config file
or production one with HTTPS. Read [documentation](./setup/README.md) file to find out how to use this variable

`GIT_USER_NAME=foo` - yangcatalog is pushing some data to GitHub repository, and it needs to
set username before any commit is done.

`GIT_USER_EMAIL=bar@foo.com` - yangcatalog is pushing some data to GitHub repository, and it needs to
set email address before any commit is done.

`CRON_MAIL_TO=bar@foo.com` - comma separated list of emails which are used
with cron jobs. If any cron job will fail it will send it to this comma separated list of email addresses.

`YANGCATALOG_CONFIG_PATH=/etc/yangcatalog/yangcatalog.conf` - Path to global config file used by all YANG Catalog components.

`CONFD_VERSION=8.0` - Version of [ConfD](https://www.tail-f.com/management-agent/) to be used.

`CONFD_PASSWORD` - Password for ConfD database.

`CONFD_PASSWORD_HASH` - Hashed password for ConfD admin in format `$1$<salt>$<hash>` or `$0$<password>`.

`CONFD_LOG=/var/yang/logs/confd` - Directory where ConfD will store log files.

`YANGLINT_VERSION=v2.1.111` - Version of [yanglint](https://github.com/CESNET/libyang) to be used.

`XYM_VERSION=0.7.0` - Version of [xym](https://github.com/xym-tool/xym) to be used.


CHANGELOG.md
## Deployment Release Notes

* ##### v5.17.0 - 2023-12-11

  * No changes - released with other [deployment submodules](https://github.com/YangCatalog/deployment)

* ##### v5.17.0 - 2023-12-11

  * No changes - released with other [deployment submodules](https://github.com/YangCatalog/deployment)

* ##### v5.16.0 - 2023-11-08

  * Improved projects setup process

* ##### v5.15.0 - 2023-10-04

  * changed Dockerfile for Opensearch [#205](https://github.com/YangCatalog/deployment/pull/205)
  * yanglint updated to 2.1.111

* ##### v5.14.0 - 2023-08-09

  * Corrected errors in documentation [#204](https://github.com/YangCatalog/deployment/pull/204)

* ##### v5.13.0 - 2023-07-06

  * Replaced elasticsearch with opensearch [#193](https://github.com/YangCatalog/deployment/issues/193)

* ##### v5.12.0 - 2023-05-31

  * Replaced api-receiver with celery-worker [#195](https://github.com/YangCatalog/deployment/issues/195)

* ##### v5.11.1 - 2023-05-03

  * No need to push yc-documentation container [#188](https://github.com/YangCatalog/deployment/issues/188)
  * Improvements to README.md file describing deployment [#185](https://github.com/YangCatalog/deployment/issues/185)

* ##### v5.11.0 - 2023-04-13

  * Rollback procedure description added to the README.md file
  * New value added to the config files

* ##### v5.10.1 - 2023-03-20

  * README.md file describing deployment to PROD created [#185](https://github.com/YangCatalog/deployment/issues/185)
  * setup.sh script and corresponding README.md file updated

* ##### v5.10.0 - 2023-03-07

  * xym tool update to version 0.7.0 [#183](https://github.com/YangCatalog/deployment/issues/183)
  * 2 new values added to the config files
  * Content of 2 new directories exposed and accessible by API now

* ##### v5.9.0 - 2023-01-26

  * Load ConfD hashed password from env file [#179](https://github.com/YangCatalog/deployment/issues/179)
  * YumaPro validator updated to version 21.10-12 [#178](https://github.com/YangCatalog/deployment/issues/178)

* ##### v5.8.0 - 2022-12-20

  * ConfD update to version 8.0 [#174](https://github.com/YangCatalog/deployment/issues/174)

* ##### v5.7.0 - 2022-11-11
  
  * sdo_analysis repository renamed to module-compilation [#154](https://github.com/YangCatalog/deployment/issues/154)
  * xym tool update to version 0.6.2
  * yanglint update to version v2.1.4 [#169](https://github.com/YangCatalog/deployment/issues/169)
  * DOCUMENTATION file reworked into the README.md for setup directory [#167](https://github.com/YangCatalog/deployment/issues/167)
  * STYLEGUIDE.md file created [#163](https://github.com/YangCatalog/deployment/issues/163)

* ##### v5.6.1 - 2022-10-10

  * xym tool update to version 0.6.1 [#161](https://github.com/YangCatalog/deployment/issues/161)

* ##### v5.6.0 - 2022-09-30

  * Frontend related repositories merged into one new repository [#154](https://github.com/YangCatalog/deployment/issues/154)
  * Overview section added to README.md file [#156](https://github.com/YangCatalog/deployment/issues/156)

* ##### v5.5.0 - 2022-08-16

   * ConfD update to version 7.8 [#149](https://github.com/YangCatalog/deployment/issues/149)
   * yanglint update to version v2.0.231 [#146](https://github.com/YangCatalog/deployment/issues/146)
   * Changes in Matomo configuration [#133](https://github.com/YangCatalog/deployment/issues/133)
   * New environment setup script created [#143](https://github.com/YangCatalog/deployment/issues/143)

* ##### v5.4.0 - 2022-07-08

  * Multiple updates to the config files [#141](https://github.com/YangCatalog/deployment/issues/141)
  * Reporting problematic IETF drafts [#139](https://github.com/YangCatalog/deployment/issues/139)
  * Default crontab mail-to address changed

* ##### v5.3.0 - 2022-06-06

  * yangcatalog.conf.sample file added [#135](https://github.com/YangCatalog/deployment/issues/135)
  * Setup Matomo for local development [#133](https://github.com/YangCatalog/deployment/issues/133)
  * Setup Kibana for local development [#130](https://github.com/YangCatalog/deployment/issues/130)
  * thesaurus.conf file updated [#129](https://github.com/YangCatalog/deployment/issues/129)
  * yanglint update to version v2.0.194 [#127](https://github.com/YangCatalog/deployment/issues/127)

* ##### v5.2.0 - 2022-05-03

  * Type checking fixes with pyright [#126](https://github.com/YangCatalog/deployment/issues/126)
  * Pyang update to version 2.5.3 [#124](https://github.com/YangCatalog/deployment/issues/124)
  * Official Slate image used for building documentation [#123](https://github.com/YangCatalog/deployment/issues/123)
  * Elasticsearch updated to version 7.10 [#120](https://github.com/YangCatalog/deployment/issues/120)
  * K8s garbage collection settings changed [#118](https://github.com/YangCatalog/deployment/issues/118)
  * Pushing more resources using HTTP/2 Push [#114](https://github.com/YangCatalog/deployment/issues/114)

* ##### v5.1.0 - 2022-03-28

  * README.md file describing NGINX config created [#117](https://github.com/YangCatalog/deployment/issues/117)
  * HTTP/2 PUSH functionality enabled [#114](https://github.com/YangCatalog/deployment/issues/114)

* ##### v5.0.0 - 2022-02-02

  * Pyang update to version 2.5.2 [#113](https://github.com/YangCatalog/deployment/issues/113)
  * Dockerfiles added to .dockerignore file
  * Docker healthchecks added for yangvalidator and yangre

* ##### v4.3.0 - 2021-12-03

  * Network between sdo-analysis and redis containers created [#110](https://github.com/YangCatalog/deployment/issues/110)
  * Use volume for Redis .rdb backup file [#109](https://github.com/YangCatalog/deployment/issues/109)
  * MariaDB completly removed from YANG Catalog [#108](https://github.com/YangCatalog/deployment/issues/108)
  * Various K8s/Docker updates and improvements in config files
  * Various healthcheck updates through solution [#107](https://github.com/YangCatalog/deployment/issues/107)

* ##### v4.2.1 - 2021-10-06


setup/setup.sh
#!/bin/bash

YANG_RESOURCES=/var/yang
cur_dir=$(pwd)

# Update and install basic apt packages
set -ex
apt update -y
apt install -y git curl rsync docker docker-compose
apt upgrade -y
apt autoremove -y

# Create a basic tree structure inside $YANG_RESOURCES directory
mkdir -p $YANG_RESOURCES
cd $YANG_RESOURCES
directories=("all_modules" "cache/redis-json" "commit_dir" "conf" "opensearch" "ietf" "ietf-exceptions" "logs/uwsgi" "logs/opensearch" "nginx" "nonietf/openconfig" "nonietf/yangmodels" "redis" "tmp" "ytrees")
for directory in ${directories[@]}; do
    mkdir -p $directory -m 755
done

# Create a basic tree structure inside $YANG_RESOURCES/nginx directory -> needed for docker volumes
cd $YANG_RESOURCES/nginx
nginx_directories=("compatibility" "drafts" "private" "results" "stats" "YANG-modules")
for directory in ${nginx_directories[@]}; do
    mkdir -p $directory -m 755
done

# Clone the directories that contain the YANG models
cd $YANG_RESOURCES/nonietf/openconfig
git clone --recurse-submodules https://github.com/openconfig/public.git
cd $YANG_RESOURCES/nonietf/yangmodels
git clone --recurse-submodules https://github.com/YangModels/yang.git

# yang-catalog@2018-04-03 module needs to be indexed to OpenSearch
cd $YANG_RESOURCES
touch yang2_repo_cache.dat
echo "{\"yang-catalog@2018-04-03/ietf\": \"/var/yang/all_modules/yang-catalog@2018-04-03.yang\"}" >yang2_repo_cache.dat

# Add iana-if-types revisions in exception file
cd $YANG_RESOURCES/ietf-exceptions
touch iana-exceptions.dat
echo "iana-if-type@2022-03-07.yang\niana-if-type@2022-08-17.yang\niana-if-type@2022-08-24.yang" >iana-exceptions.dat

# Make sure module yang-catalog@2018-04-03 is available
cd $YANG_RESOURCES/all_modules
curl -X GET https://raw.githubusercontent.com/YangModels/yang/main/experimental/ietf-extracted-YANG-modules/yang-catalog%402018-04-03.yang -o yang-catalog@2018-04-03.yang
# Store file to the cache/redis-json to be able to load it at the start of yc-api-recovery
cd $cur_dir
cp yang-catalog.json $YANG_RESOURCES/cache/redis-json/$(date +"%Y-%m-%d_00:00:00-UTC.json")
cp ../conf/redis_databases.json $YANG_RESOURCES/redis/redis_databases.json
cp ../conf/yangcatalog.conf.sample $YANG_RESOURCES/conf/yangcatalog.conf
# Add yang user -> whole $YANG_RESOURCES tree structure needs to belong to user 'yang'
groupadd -g 1001 -r yang
useradd -r -g yang -u 1016 yang
chown -R yang:yang $YANG_RESOURCES

# Pull RFCs and Draft files needed for module-compilation (this might take some time)
cd $YANG_RESOURCES/ietf
rsync -avz --include 'draft-*.txt' --include 'draft-*.xml' --exclude '*' --delete rsync.ietf.org::internet-drafts my-id-mirror
# rsync -avz --include 'draft-*.txt' --include 'draft-*.xml' --exclude '*' --delete rsync.ietf.org::id-archive my-id-archive-mirror
rsync -avlz --delete --include="rfc[0-9]*.txt" --exclude="*" ftp.rfc-editor.org::rfcs rfc

chmod 755 $YANG_RESOURCES

chown -R yang:yang $YANG_RESOURCES
